File: sqlite_test.py
import sqlite3
import uuid
import time

# Определяем класс для хранения данных
class DataObject:
    def __init__(self, name, value):
        self.name = name
        self.value = value

# Функция для создания таблицы в базе данных
def create_table(cursor):
    cursor.execute('''CREATE TABLE IF NOT EXISTS data (
                        name TEXT,
                        value TEXT
                    )''')
    cursor.execute("DELETE FROM data")

# Функция для генерации данных
def generate_data(num_objects):
    data_objects = []
    for _ in range(num_objects):
        name = str(uuid.uuid4())  # Генерация случайного UUID
        value = str(uuid.uuid4())
        data_objects.append(DataObject(name, value))
    return data_objects

# Функция для вставки данных в базу данных
def insert_data(conn, cursor, data_objects):
    start_time = time.time()
    for obj in data_objects:
        cursor.execute("INSERT INTO data (name, value) VALUES (?, ?)", (obj.name, obj.value))
    conn.commit()
    end_time = time.time()
    return end_time - start_time
# Функция для выполнения SQL-запроса с измерением времени
def execute_query(conn,cursor, query):
    start_time = time.time()
    res = cursor.execute(query)
    res.fetchall()
    end_time = time.time()
    return end_time - start_time

# Основная функция
def main():
    # Подключение к базе данных
    conn = sqlite3.connect('History')
    cursor = conn.cursor()

    # Создание таблицы, если её ещё нет
    create_table(cursor)

    # Генерация данных
    data_objects = generate_data(50000)

    # Вставка данных в базу
    execution_time = insert_data(conn, cursor, data_objects)
    print(f"Time taken for query execution: {execution_time} seconds")

    # Выполнение запросов
    queries = [
        "SELECT * FROM data",
    ]
    for query in queries:
        execution_time = execute_query(conn, cursor, query)
        print(f"Time taken for query execution: {execution_time} seconds")

    conn.close()

if __name__ == "__main__":
    main()

File: sqlite_test/sqlite_test.cpp
﻿#include <iostream>
#include "TestObj.h"
#include "sqlite3.h"
#include <windows.h>
#include <string> 
#include "Timer.h"
using namespace std;

// Функция обратного вызова для вывода данных из запроса
static int my_special_callback(void* unused, int count, char** data, char** columns)
{
    int idx;

    for (idx = 0; idx < count; idx++) {
        printf("The data in column \"%s\" is: %s\n", columns[idx], data[idx]);
    }

    printf("\n");

    return 0;
}

// Генерация нового GUID
static string GetNewGUID()
{
    UUID uuid = { 0 };
    string guid;

    // Создаем uuid
    UuidCreate(&uuid);

    // Конвертируем uuid в строку
    RPC_CSTR szUuid = NULL;
    if (UuidToStringA(&uuid, &szUuid) == RPC_S_OK)
    {
        guid = (char*)szUuid;
        RpcStringFreeA(&szUuid);
    }

    return guid;
}

// Открытие БД SQLite
static sqlite3* OpenDatabase()
{
    sqlite3* Database;
    int openResult = sqlite3_open16(L"../History", &Database);
    if (openResult != SQLITE_OK) {
        wcout << L"Ошибка открытия БД" << endl;
    }
    return Database;
}

// Создание таблицы
static void CreateTable(sqlite3* Database)
{
    char* errmsg;
    char sql[] = "CREATE TABLE IF NOT EXISTS test("
        "name TEXT NOT NULL,"
        "value TEXT NOT NULL);";
    // Создаем таблицу
    int execResult = sqlite3_exec(Database, sql, NULL, NULL, &errmsg);
    if (execResult != SQLITE_OK) {
        cout << errmsg << endl;
        wcout << L"Ошибка выполнения запроса" << endl;
    }
}
static void ExecuteQuery(sqlite3* Database, const char* query)
{
    char* errmsg;
    Timer timer;

    sqlite3_stmt* pStatement;
    int execResult = sqlite3_prepare_v2(Database, query, -1, &pStatement, NULL);
    if (execResult != SQLITE_OK) {
        wcout << L"Ошибка подготовки запроса" << endl;
        return;
    }

    timer.start();
    int step = sqlite3_step(pStatement);
    if (step == SQLITE_ROW) {
        const unsigned char* n = sqlite3_column_text(pStatement, 0);
        wcout << L"Получены следующие данные " << *n  << endl;
    }
    sqlite3_reset(pStatement);
    timer.stop();

    std::cout << "Query time: "<< query << std::endl;
    std::cout << "Seconds: " << timer.elapsedSeconds() << std::endl;
}
static void InsertData(sqlite3* Database)
{
    // Генерируем и вставляем тестовые данные
    TestObj* rows = new TestObj[50000];
    for (int i = 0; i < 50000; i++) {
        rows[i] = TestObj();
    }

    Timer timer;
    sqlite3_stmt* res;  // компилируемое выражение
    int rc = sqlite3_prepare_v2(Database, "INSERT INTO test (name, value) VALUES (?, ?)", -1, &res, 0);

    timer.start();
    sqlite3_exec(Database, "BEGIN;", nullptr, nullptr, nullptr);
    for (int i = 0; i < 50000; i++) {
        sqlite3_bind_text(res, 1, rows[i].Name, -1, SQLITE_STATIC);
        sqlite3_bind_text(res, 2, rows[i].Value, -1, SQLITE_STATIC);
        // выполняем выражение
        int step = sqlite3_step(res);
        // если выражение успешно выполнено
        if (step == SQLITE_ERROR)
        {
            printf("Error\n");
        }
        sqlite3_clear_bindings(res);
        sqlite3_reset(res);
    }
    sqlite3_exec(Database, "COMMIT;", nullptr, nullptr, nullptr);
    timer.stop();

    std::cout << "Insert time:" << std::endl;
    std::cout << "Seconds: " << timer.elapsedSeconds() << std::endl;
    delete[] rows;
}
int main()
{
    setlocale(LC_ALL, "Russian");

    // Открываем БД
    sqlite3* Database = OpenDatabase();

    // Создаем таблицу
    CreateTable(Database);

    InsertData(Database);

    // Выполняем запросы и измеряем время выполнения
    ExecuteQuery(Database, "SELECT * FROM test");

    // Удаляем данные из таблицы
    char* errmsg;
    int execResult6 = sqlite3_exec(Database, "DELETE FROM test", NULL, NULL, &errmsg);
    if (execResult6 != SQLITE_OK) {
        cout << errmsg << endl;
        wcout << L"Ошибка выполнения запроса" << endl;
    }

    // Закрываем соединение с БД
    sqlite3_close(Database);
    return 0;
}